client/src/main/java/com/tencent/rss/client/impl/ShuffleWriteClientImpl.java

```diff
@@ -156,41 +172,72 @@ public SendShuffleDataResult sendShuffleData(String appId, List<ShuffleBlockInfo
-        }
-        serverToBlockIds.get(ssi).add(sbi.getBlockId());
 
-        if (!serverToBlocks.containsKey(ssi)) {
-          serverToBlocks.put(ssi, Maps.newHashMap());
-        }
-        Map<Integer, Map<Integer, List<ShuffleBlockInfo>>> shuffleIdToBlocks = serverToBlocks.get(ssi);
-        if (!shuffleIdToBlocks.containsKey(shuffleId)) {
-          shuffleIdToBlocks.put(shuffleId, Maps.newHashMap());
-        }
-
-        Map<Integer, List<ShuffleBlockInfo>> partitionToBlocks = shuffleIdToBlocks.get(shuffleId);
-        if (!partitionToBlocks.containsKey(partitionId)) {
-          partitionToBlocks.put(partitionId, Lists.newArrayList());
-        }
-        partitionToBlocks.get(partitionId).add(sbi);
+    // In order to reduce the data to send in quorum protocol,
+    // we split these blocks into two rounds: primary and secondary.
+    // The primary round contains [0, replicaWrite) replicas,
+    // which is minimum number when there is no sending server failures.
+    // The secondary round contains [replicaWrite, replica) replicas,
+    // which is minimum number when there is at most *replicaWrite - replica* sending server failures.
+    for (ShuffleBlockInfo sbi : shuffleBlockInfoList) {
+      List<ShuffleServerInfo> allServers = sbi.getShuffleServerInfos();
+      if (replicaSkip) {
+        genServerToBlocks(sbi, allServers.subList(0, replicaWrite),
+          primaryServerToBlocks, primaryServerToBlockIds);
+        genServerToBlocks(sbi, allServers.subList(replicaWrite, replica),
+          secondaryServerToBlocks, secondaryServerToBlockIds);
+      } else {
+        // When replicaSkip is disabled, we send data to all replicas within one round.
+        genServerToBlocks(sbi, allServers,
+          primaryServerToBlocks, primaryServerToBlockIds);
       }
     }
 
+    // maintain the count of blocks that have been sent to the server
+    Map<Long, AtomicInteger> blockIdsTracker = Maps.newConcurrentMap();
+    primaryServerToBlockIds.values().forEach(
+      blockList -> blockList.forEach(block -> blockIdsTracker.put(block, new AtomicInteger(0)))
+    );
+    secondaryServerToBlockIds.values().forEach(
+      blockList -> blockList.forEach(block -> blockIdsTracker.put(block, new AtomicInteger(0)))
+    );
+
     Set<Long> failedBlockIds = Sets.newConcurrentHashSet();
     Set<Long> successBlockIds = Sets.newConcurrentHashSet();
     // if send block failed, the task will fail
     // todo: better to have fallback solution when send to multiple servers
-    sendShuffleDataAsync(appId, serverToBlocks, serverToBlockIds, successBlockIds, failedBlockIds);
+
+    // sent the primary round of blocks.
+    boolean isAllSuccess = sendShuffleDataAsync(
+        appId, primaryServerToBlocks, primaryServerToBlockIds, blockIdsTracker);
+
+    // The secondary round of blocks is sent only when the primary group issues failed sending.
+    // This should be infrequent.
+    if (!isAllSuccess && replicaSkip) {
```

Comment:
`replicaSkip` is not necessary.

Reply:
This is the new option

Reply:
If replicaSkip is enable, collection is empty.We won't send any rpc.It's up to you.

Reply:
Ok, I will use `!primaryServerToBlocks.isEmpty()` instead.

Reply:
SendaryServerToBlocks.isEmpty()?

Reply:
Yes, copy error...